File: client/src/communication/packet.go
package communication

import (
	"fmt"
	"strings"
)

const newLine = "\n"
const headerContentSeparator = "\n\n"
const tagValueSeparator = ": "

type Packet struct {
	Command   string
	ClientId  string
	MessageId string
	Content   string
}

func (p *Packet) Serialize() string {
	command := fmt.Sprintf("COMMAND%s%s%s", tagValueSeparator, p.Command, newLine)
	clientId := fmt.Sprintf("CLIENT_ID%s%s%s", tagValueSeparator, p.ClientId, newLine)
	messageId := fmt.Sprintf("MESSAGE_ID%s%s%s", tagValueSeparator, p.MessageId, newLine)
	header := command + clientId + messageId
	content := p.Content
	packet := header + newLine + content
	return packet
}

type PacketResponse struct {
	Status  string
	Content string
}

func DeserializePacketResponse(data string) *PacketResponse {
	response := strings.Split(data, headerContentSeparator)
	headers := response[0]
	content := response[1]
	status := ""
	for _, header := range strings.Split(headers, newLine) {
		line := strings.Split(header, tagValueSeparator)
		tag := line[0]
		value := line[1]
		if tag == "STATUS" {
			status = value
		}
	}
	return &PacketResponse{Status: status, Content: content}
}

File: server/src/communication/packet.py
class Packet:
    _tag_value_separator = ": "
    _new_line_separator = "\n"
    _header_content_separator = "\n\n"

    def __init__(self, command, client_id, message_id, data):
        self.command = command
        self.client_id = client_id
        self.message_id = message_id
        self.data = data

    def deserialize(serializedPacket):
        occurrences = 1
        header, content = serializedPacket.split(Packet._header_content_separator, occurrences)
        header_options = header.split(Packet._new_line_separator)

        command = None
        client_id = None
        message_id = None
        for option in header_options:
            tag, value = option.split(Packet._tag_value_separator, occurrences)
            if tag == "COMMAND":
                command = value
            elif tag == "CLIENT_ID":
                client_id = value
            elif tag == "MESSAGE_ID":
                message_id = value

        return Packet(command, client_id, message_id, content)

class PacketResponse:
    status_ok = "OK"
    status_ko = "KO"

    def __init__(self, status, data):
        self.status = status
        self.data = data

    def serialize(packetResponse):
        return f"STATUS: {packetResponse.status}\n\n{packetResponse.data}"
